### ISSUE TITLE
LM3/4 - HotSpot Levelling

### ISSUE THREAD

**Issue by UltimaOath** [ArcanePlugins/LevelledMobs#447]
Initial concept in Discord:  
https://discord.com/channels/752310043214479462/756850940694495253/1080675231309500436

```yaml
# This system combines the kills of all players together for simplicity sake; while I like the
# idea of doing this per-player it would add an unncessary layer of complexity to manage the
# chunks on a per-player basis.
# All entities initially start at Level 1 (or the minLevel set for the scenario).

presets:
  hotspot_Levelling:
    # This Strategy Preset controls the player-variable based level modification system.
    name: 'LVLling Strategy - Hotspot'
    strategies:
      hotspot-levelling:
      # The Level Multiplier will multiply the final value by this amount, providing the scale
      # to how the levels are adjusted. A lower multiplier will stretch the scale out, meaning
      # it would take more entities to increase the level.
        level-multiplier: 1.25
      # The Level Modifier will add an amount to the final value prior to applying the multiplier.
      # This helps to provide a strength component to the scale. A higher modifier will
      # add value to the average level before it's applied to new entities.
        level-modifier: 2.0
      # This cooldown will apply to each entity killed, and the entities will be dropped from
      # the list once the cooldown has been reached based on that entities' killtime.
        cooldown-timer: 600s  # 10 minutes
      # This represents any nearby chunks that are affected by the new level average; if a chunk
      # is already set to a different level or has not been set yet, the highest level available
      # between the three will be used. If this is set to zero, then setting not used.
        apply-to-nearby-chunks: 5
```

The idea is that the world is level 1/minlevel, player(s) go to kill entities; their levels, the chunk they died in, and time of death are recorded in a list. A secondary list is made and sorted into the different affected chunks. Those chunks are given an 'average level value' which is generated by taking 'a sum of the levels of all entities which died in the same chunk' and 'a count of all entities which were used in the sum', then applying the modifier and multiplier settings will begin to increase the average over time based on the number of entities killed and the level of those entities being taken into account.   
This map of chunks with their level averages set is what sets the level of any subsequently spawned in entities within the affected chunks. These higher level entities will help to boost the average faster if players are determined to stay in the area, and if a player can handle it they are rewarded with the higher level entities. If they can't they can leave the area to an unaffected area of chunks or to an area with a lower average at the time until things calm down in the area they were working in. Using the 'nearby chunks' feature helps to spread the effect out away from the player as they travel around.

Once an entities' killtime has exceeded the cooldown timer, that entity is dropped from the list and therefore it is no longer affecting the averages. There are two effects to this: if the player has left or stopped fighting then the average will begin to reduce. If the player remains and is still fighting it out, then it is only adding fuel to the fire as the lower levels are removed from the averages and are replaced with higher levels. (edited)  
\[7:39 PM\]  
I could also see this being a lvl-mod\_ rather than an actual strategy; it would override whatever level was set in a strategy by changing the levels of future spawned entities only in the affected chunks, but leaving the unaffected chunks to the original strategy.

A sample/demonstration excel spreadsheet regarding this system is [available here](https://cdn.discordapp.com/attachments/756850940694495253/1080675230999126037/HotSpotLevelling.xlsx).

**Comment by UltimaOath** [comment]
Got bored after waking up and decided randomly to sudocode this idea; I also had a few modifications to the general concept mostly associated with converting the entire thing into three separate level modifiers rather than true levelling strategies.  


```yaml
# This system combines the kills of all players together for simplicity sake; while I like the
# idea of doing this per-player it would add an unnecessary layer of complexity to manage the
# chunks on a per-player basis.

presets:
  lvl-mod_Hotspots:
    # This Level Modifier covers the 'Hotspots' mechanic.
    # HOTSPOTS_MOD = ( ( SUM-ENTITY-LEVELS ) / ( ENTITY-COUNT ) ) + LVL-MOD ) x LVL-MULTI )
    name: 'LVL Modifier - Hotspots'
    strategies:
      hotspots_Modifier:
      # The hotspots penalty will add the specified amount to the final hotspots mod value prior
      # to applying the hotspots multiplier. This adds additional strength to future levels being
      # applied. A higher value will cause greater leaps in strength over time.
        hotspots-penalty: 2.0
      # The hotspots multiplier will multiply the specified amount to the final hotspots value after
      # the previous hotspot penalty. This is used to stretch or shrink the scale of how many more
      # entities it would take to jump to the next difficulty tier. A lower value will stretch the scale
      # while a higher value would shrink the scale exponentially. 
        hotspots-multiplier: 1.25
      # This is the highest value (cap) the hotspot modifier can apply to a level.
        hotspots-cap: 10
      # This cooldown will apply to each entity killed, and the entities will be dropped from
      # the list once the cooldown has been reached based on that entities' killtime.
        hotspots-cooldown: 10m
      # This will apply any affects of this modifier to all surrounding chunks within this distance.
      # Each of these modifiers are unique, so would stack on to each other.
        apply-to-nearby-chunks: 5


  lvl-mod_Barricades:
    # This Level Modifier covers the 'Barricades' mechanic.
    # BARRICADES_MOD = ( BARRICADE-PENALTY + BARRICADE-REGEN )
    name: 'LVL Modifier - Barricades'
    strategies:
      barricades_Modifier:
      # The barricades penalty will add the specified value for each elapsed timer instance, up to the cap.
        barricades-penalty: 0.25
      # The barricades cooldown regen will remove the specified value for each elapsed timer instance, until the modifier is zeroed out.
        barricades-cooldown-regen: '-0.5'
      # This is the highest value (cap) the barricade modifier can apply to a level.
        barricades-cap: 10
      # This is the timer by which the penalty and cooldown are measured and increased/decreased.
        barricades-timer: 5m
      # This cooldown will drop players from the list based on the timestamp of the players' exit.
        barricades-cooldown: 15m
      # This will apply any affects of this modifier to all surrounding chunks within this distance.
      # Each of these modifiers are unique, so would stack on to each other.
        apply-to-nearby-chunks: 5


  lvl-mod_Heatmaps:
    # This Level Modifier covers the 'Heatmaps' mechanic.
    # HEATMAPS_MOD = ( HEATMAPS-PENALTY + HEATMAPS-REGEN )
    name: 'LVL Modifier - Heatmaps'
    strategies:
      heatmaps_Modifier:
      # The heatmap penalty will add the specified value for each elapsed timer instance, up to the cap.
        heatmaps-penalty: 0.25
      # The barricades cooldown regen will remove the specified value for each elapsed cooldown instance, until the modifier is zeroed out.
        heatmaps-cooldown-regen: '-1.0'
      # This is the highest value (cap) the barricade modifier can apply to a level.
        heatmaps-cap: 5
      # This cooldown will drop players from the list based on the timestamp of the players.
        heatmap-cooldown: 15m
      # This will continue to add players to the chunk even if they are already listed.
        recount-players: true
      # This will apply any affects of this modifier to all surrounding chunks within this distance.
      # Each of these modifiers are unique, so would stack on to each other.
        apply-to-nearby-chunks: 5
```


Hotspot Modifier: 
TLDR Meant to increase the difficulty of an area of above-average combat.

When a player kills an entity, a log is generated which maintains the 'master list'. In this master list, we check if the chunk in which the entity was killed is already in the list and if so we add to a nested list within that chunk the id of the entity, it's level, and the timestamp of it's death. If the chunk which the entity was killed in isn't part of the list, a new entry is created for that chunk following the nested list format previously mentioned. Entities are dropped from the list after `hotspots-cooldown:` duration elapses, and the modifier can only reach as high as the `hotspots-cap:` . 
Using the formula `HOTSPOTS_MOD = ( ( SUM-ENTITY-LEVELS ) / ( ENTITY-COUNT ) ) + LVL-PENALTY ) x HOTSPOTS-MULTIPLIER )`, when we need to generate a modifier value for any newly summoned entities' level, we first need to take the chunk the entity is spawned in and reference the master list. If the chunk specified isn't on the list, the entity isn't to be modified by this system and moves on. If it is, then we need to take the sum of all levels for all entities within the list and then divide that value by the total number of entities killed in that chunk who remain on the master list. From here we need to add the hotspots-penalty to the value, and then apply the hotspots-multiplier to the next to receive the final hotspots modifier.


Barricades Modifier:
TLDR Meant to increase the difficulty of an area where players are encamped. 

When a player enters a chunk, a log is generated which maintains a 'master list'. In this master list, we check if the chunk the player has entered is already listed and if so we add to the nested list within that chunk the players' id and the timestamp associated with the last time the player had entered or exited the chunk. When the `barricades-timer:` pings, the master list is grabbed and all chunks within the list are checked. If the inspected chunk contains players, those players are added to the master list. Then for each chunk in the list, you check for any player that is listed more than once within the check (this is because any entity that was in the chunk when first pinged via entrance or prior ping, should it still be in the chunk once the timer had elapsed, then that means they either never left or didn't go far; therefore barricading themselves ). For each instance that a player is listed multiple times, you increase the barricade penalty by the configured amount up to the cap. When the barricade cooldown timer elapses, the master list is updated and pulled, and for each chunk any player whose timestamp has elapsed the cooldown has their entry dropped from the list. If a player is dropped from a specified chunk, the barricade regen is increased by the configured amount. The final modifier can only be zero (not applied) or up to the cap.


Heatmap Levelling Modifier:
TLDR Meant to increase the difficulty of an area where players often travel/pass through. 

When a player enters a chunk, a log is generated which maintains the 'master list'. This master list needs to have each chunk listed, and nested within each chunk a log of each instance of a player entering the specified chunk with an associated timestamp of the instance. If `recount-players: true` , then a player can have multiple instances placed within the chunk should they enter, exit, then reenter the chunk again; with this setting `false` then if a player is already associated with that chunk they will not be added a further time. Each timestamped entry will be dropped from the list after the `heatmap-cooldown:` duration has elapsed.

**Comment by UltimaOath** [comment]
Updated my last post to make the processing less taxing, absorbing 'heatmaps' into the 'barricades' system, and gave a stronger explaination of the backend.

**Comment by UltimaOath** [comment]
Suggested to shift formatting into `modifiers` section as subsection of `strategies`.  
  
Example:  
  
```yaml
presets:
  lvl-mod_Hotspots:
    name: 'LVL Modifier - Hotspots'
    strategies:
      modifiers:
        hotspots_Modifier:
          hotspots-penalty: 2.0
          hotspots-multiplier: 1.25
          hotspots-cap: 10
          hotspots-cooldown: 10m
          apply-to-nearby-chunks: 5
          hotspots-placeholder: hotspots_mod  #would create placeholder `%hotspots_mod%` for use


  lvl-mod_Barricades:
    name: 'LVL Modifier - Barricades'
    strategies:
      modifiers:
        barricades_Modifier:
          barricades-penalty: 0.25
          barricades-cap: 10
          barricades-timer: 45s
          barricades-cooldown: 5m
          apply-to-nearby-chunks: 5
          barricades-placeholder: barricades_mod  #would create placeholder `%barricades_mod%` for use
```
